cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/replication/GetRevisionTaskThreaded.java

```diff
@@ -137,14 +142,36 @@ public boolean add(Future<DocumentRevsList> documentRevsListFuture) {
         // we make the request at construction time...
         for (final BulkGetRequest request : requests) {
 
-            completionService.submit(new Callable<DocumentRevsList>() {
+            submittedJobs.add(completionService.submit(new Callable<DocumentRevsList>() {
                 @Override
                 public DocumentRevsList call() throws Exception {
                     return new DocumentRevsList(GetRevisionTaskThreaded.this.sourceDb.getRevisions
                             (request.id, request.revs, request.atts_since,
                                     GetRevisionTaskThreaded.this.pullAttachmentsInline));
                 }
-            });
+            }));
+        }
+    }
+
+    void cancel() {
+        synchronized (submittedJobs) {
+            for (Future<DocumentRevsList> f : submittedJobs) {
+                // Don't interrupt if already running, we'll get the running ones from the completed
+                // queue.
+                f.cancel(false);
+            }
+            // Cancelling will line a whole bunch of tasks up for addition to the queue, so we need
+            // to keep draining it
+            try {
+                while (!submittedJobs.isEmpty()) {
+                    Future<DocumentRevsList> r = completionService.take();
```

Comment:
I've changed `take()` to `poll()` in 8223c28, but in that case if there is a race condition then the while loop might remain active polling the queue (in which case poll with timeout might actually be preferable).
I'm reasonably confident that the synchronized set should prevent any race condition, unless we've missed another code branch where `submittedJobs.remove()` needs to be called.
wdyt?